app/models/course/condition/achievement.rb

```diff
@@ -3,4 +3,50 @@ class Course::Condition::Achievement < ActiveRecord::Base
   belongs_to :achievement, class_name: Course::Achievement.name, inverse_of: false
 
   default_scope { includes(:achievement) }
+
+  validate :validate_achievement_condition, if: :achievement_id_changed?
+
+  delegate :title, to: :achievement
+
+  private
+
+  # Given an achievement, returns all its achievement conditions.
+  #
+  # @param [Course::Achievement] achievement The achievement to return achievement conditions for.
+  # @return [Array<Course::Achievement>]
+  def achievement_conditions(achievement)
+    # Course::Condition::Achievement.
+    #   joins { condition.conditional(Course::Achievement) }.
+    #   where { condition.conditional.id == achievement.id }.
+    #   map(&:achievement)
+
+    # Workaround, pending the squeel bugfix (activerecord-hackery/squeel#390) that will allow
+    # allow the above query to work without #reload
+    Course::Achievement.joins(
+      "INNER JOIN
+        (SELECT cca.achievement_id
+          FROM course_condition_achievements cca INNER JOIN course_conditions cc
+            ON cc.actable_type = 'Course::Condition::Achievement' AND cc.actable_id = cca.id
+            WHERE cc.conditional_id = #{achievement.id}) ids
+    ON ids.achievement_id = course_achievements.id")
+  end
+
+  def validate_achievement_condition
+    validate_references_self
+    validate_unique_dependency
+  end
+
+  def validate_references_self
+    errors.add(:achievement,
+               I18n.t('activerecord.errors.models.achievement_condition.'\
```

Comment:
You should probably not use I18n.t directly. Use `ActiveModel::Errors.add`, i.e.

```ruby
errors.add(:achievement, :references_self)
```

then in the translation

```
en:
  activerecord:
    errors:
      models:
        course/condition/achievement:
          attributes:
            achievement:
              unique_dependency: 'achievement cannot have duplicate conditions'
              references_self: 'achievement cannot have itself as a condition'
```

Maybe the i18n structure for models (within the activerecord namespace) should just mirror the models directory. thoughts?